### ISSUE TITLE
OSSL_PARAM_BLD_push_BN of zero secure BIGNUM leads to an OSSL_PARAM with data=NULL and data_size=0

### ISSUE THREAD

**Issue by ifranzki** [openssl/openssl#20011]
Due to a bug in my own provider, I encountered a strange situation. Not sure if its a bug or works as designed though....

Within my provider export function, I used `OSSL_PARAM_BLD_push_BN()` to push a BIGNUM that was allocated via `BN_secure_new()`, but was never actually set a value (due to a bug in my code), so its value was zero. While additional parameters were pushed to the param-builder, this one was the only with a secure BIGNUM. 

This caused an address sanitizer error when importing these parameters into another provider. The other provider was using `OSSL_PARAM_get_BN` to get the BIGNUM of that parameter.  The error finally occurred in `bin2bn()` (called via `BN_native2bn()` -> `BN_bin2bn()` -> `bin2bn()`) because parameter `s` was NULL and `len` was also zero:

```
    if (endianess == LITTLE) {
        s2 = s + len - 1;
        inc2 = -1;
        inc = 1;
    } else {
        s2 = s;
        inc2 = 1;
        inc = -1;
        s += len - 1;  <----- here we add -1 to NULL which was detected by the address sanitizer
    }
```

It turned out that the OSSL_PARAM created via `OSSL_PARAM_BLD_to_param()` has `data = NULL` and `data_size = 0`. There is no check in `OSSL_PARAM_get_BN()` before passing `data` to `BN_native2bn(`) and also `bin2bn()` does not check if the buffer `s` is NULL or `len` is zero.

However, generating an OSSL_PARAM with a NULL data field by `OSSL_PARAM_BLD_to_param()` is already wrong. 
What happens is the following:

`OSSL_PARAM_BLD_push_BN()` calls `push_BN()` with parameter `sz = NULL`, because `BN_num_bytes(bn)` results in zero bytes for a zero BIGNUM. `push_BN()` then calls `param_push()` again with the zero `sz` and zero `alloc` parameters. Consequently, the `pd->alloc_blocks` is also zero, and it adds zero to `bld->secure_blocks`. If this is the one and only secure parameter, then  `bld->secure_blocks` remains zero. 

`OSSL_PARAM_BLD_to_param()` only allocates a secure buffer for the builder if `bld->secure_blocks` is non-zero. In my case, it is zero, and thus `s` stays NULL.

`param_bld_convert()` then converts the parameters, and uses the passed `secure` pool (which is NULL) to obtain the address of the buffer for the secure parameter:

```
        if (pd->secure) {
            p = secure;
            secure += pd->alloc_blocks;
        } 
```
Since `secure` is NULL, `p` is also NULL. 

The following `BN_bn2nativepad()` would attempt to write to `p`, but since the BIGNUM is zero, it does not output any bytes, and thus ignores that the `to` buffer is NULL (`tolen` is also zero).

As a result we have an OSSL_PARAM with `data = NULL` and `data_size = 0`.

I am aware that it is very unlikely to pass a zero secure BIGNUM via param builder, but maybe its worth to fix this anyway.....
I have no idea how to fix this though, since a zero BIGNUM obviously occupies zero bytes. Maybe we need to special-case this and force it to produce a 1 zero byte OSSL_PARAM for a zero BIGNUM.....

**Comment by paulidale** [comment]
Allocating 1 byte is fix I thought of while reading it.

**Comment by slontis** [comment]
This sounds like an error case to me, not a silent pass..

**Comment by levitte** [comment]
I agree with the 1 byte idea.

**Comment by slontis** [comment]
This hides the actual problem, which was unintentionally not setting the BIGNUM.

**Comment by levitte** [comment]
> This hides the actual problem, which was unintentionally not setting the BIGNUM.

Well, not quite.  A new BIGNUM is always a zero.  There's not such thing, conceptually speaking, as an unset BIGNUM.

What if a zero was intentional?  In that case, the issue presented here would still be a bug, would it not?

Besides, we push a positive amount of bytes (more than one, incidently) with all other numeric types, even when the value is zero, so pushing nothing for BIGNUMs specifically is inconsistent.

**Comment by slontis** [comment]
I was referring to this comment..
````It turned out that the OSSL_PARAM created via OSSL_PARAM_BLD_to_param() has data = NULL and data_size = 0.````

**Comment by levitte** [comment]
Ah ok

**Comment by levitte** [comment]
Interestingly enough, if you have other secure params, the `data` OSSL_PARAM field will not be NULL (`data_size` will still be zero).  That should work transparently (`bin2bn` will simply not read any bytes at all from the input pointer).

This bug turns out to be more subtle than I expected...

**Comment by slontis** [comment]
(I see what you mean after running the code locally)..
Just doing bn = BN_new() behaves differently from bn = BN_secure_new() (The secure case does set NULL, 0)

**Comment by levitte** [comment]
@ifranzki, would you mind testing your provider against an OpenSSL that's patched with #20013?
